src/Std/Data/DHashMap/Internal/RawLemmas.lean

```diff
@@ -832,13 +833,28 @@ theorem contains_keys [EquivBEq α] [LawfulHashable α] (h : m.1.WF) {k : α} :
 @[simp]
 theorem mem_keys [LawfulBEq α] [LawfulHashable α] (h : m.1.WF) {k : α} :
     k ∈ m.1.keys ↔ m.contains k := by
-  simp_to_model 
-  rw [List.containsKey_eq_keys_contains]
+  simp_to_model
+  rw [List.containsKey_eq_keys_contains, List.elem_iff]
 
 theorem distinct_keys [EquivBEq α] [LawfulHashable α] (h : m.1.WF) :
     m.1.keys.Pairwise (fun a b => (a == b) = false) := by
   simp_to_model using (Raw.WF.out h).distinct.distinct
 
+@[simp]
+theorem toList_map_fst {α β} (m : Raw₀ α β) :
+    m.1.toList.map Sigma.fst = m.1.keys := by
+  simp_to_model
+  simp [List.keys_eq_map_fst]
+
+open _root_.List in
+theorem toList_insert_perm_of_not_contains [EquivBEq α] [LawfulHashable α] (h : m.1.WF)
+    (k : α) (v : β k) (h' : m.contains k = false) :
+    (m.insert k v).1.toList ~ ⟨k, v⟩ :: m.1.toList := by
+  revert h'
+  simp_to_model
+  intro h'
+  exact Perm.trans (toListModel_insert (Raw.WF.out h)) (by simp [insertEntry, h'])
```

Comment:
Here is how I had envisioned this working: you prove
```lean
private theorem perm_of_perm {l₁ l₂ l₃ : List α} (h : List.Perm l₁ l₂) :
    List.Perm l₁ l₃ ↔ List.Perm l₂ l₃ :=
  ⟨h.symm.trans, h.trans⟩
```
and add ``← `(perm_of_perm)`` to `congrNames`. Then this lemma is proved like this:
```suggestion
open _root_.List in
theorem toList_insert_perm_of_not_contains [EquivBEq α] [LawfulHashable α] (h : m.1.WF)
    (k : α) (v : β k) : m.contains k = false →
    (m.insert k v).1.toList ~ ⟨k, v⟩ :: m.1.toList := by
  simp_to_model using Perm.of_eq ∘ insertEntry_of_containsKey_eq_false
```

To spell it out explicitly: after `simp_to_model`, it should always be possible to `generalize toListModel m.1.buckets = l` so that `m` is then no longer present in the goal.